earthaccess/auth.py

```diff
@@ -21,6 +24,15 @@
 
 
 logger = logging.getLogger(__name__)
+CLIENT_ID = "ntD0YGC_SM3Bjs-Tnxd7bg"
+
+# Host URL options, for different Earthdata domains.
+Environment_Definitions = namedtuple("Environment_Definitions", ["PROD", "UAT", "SIT"])
+Env = Environment_Definitions(
+    {"edl": "urs.earthdata.nasa.gov", "cmr": CMR_OPS},
+    {"edl": "uat.urs.earthdata.nasa.gov", "cmr": CMR_UAT},
+    {"edl": "sit.urs.earthdata.nasa.gov", "cmr": CMR_SIT},
+)
```

Comment:
IIRC, part of the discussion we had a little while back was to find a good name to represent "environment" that avoids confusion with the "environment" login strategy.

I think the code above may still has potential for confusion, and has also led to a rather long parameter name elsewhere (`earthdata_environment`). To address this, and also to introduce some better typing constructs, here's what I suggest in place of the above, which uses the term "system" rather than "environment":

```python
from dataclasses import dataclass

from typing_extensions import NewType

import cmr

CMRBaseURL = NewType("CMRBaseURL", str)
EDLHostname = NewType("EDLHostname", str)


@dataclass(frozen=True)
class System:
    cmr_base_url: CMRBaseURL
    edl_hostname: EDLHostname


PROD = System(CMRBaseURL(cmr.CMR_OPS), EDLHostname("urs.earthdata.nasa.gov"))
UAT = System(CMRBaseURL(cmr.CMR_UAT), EDLHostname("uat.urs.earthdata.nasa.gov"))
SIT = System(CMRBaseURL(cmr.CMR_SIT), EDLHostname("sit.urs.earthdata.nasa.gov"))
```

I also suggest that this code be placed in its own `system.py` file, as I don't feel that putting this in `auth.py` is quite the right spot. Yes, the EDL hostname is related to auth, but making use of a "system" does not _require_ authorization.

For example, when calling the `api.search_data` function, we might want to specify a `mode` argument, like so:

```python
search_data(mode=UAT.cmr_base_url, ...)
```

Ignoring the questionable use of the name `mode` (from `python_cmr`), I can specify a different "mode" without requiring auth, so to have to import "system" from the `auth` module is perhaps not particularly intuitive.

Perhaps @betolink, @mfisher87, and @jhkennedy can provide some feedback on this suggestion before proceeding.

Reply:
The code you propose and suggest putting in `system.py` looks fine to me!

Maybe we want a simpler API?  

Such as case-insensitive strings, with acceptable options of 'UAT', 'SIT', and 'PROD' — for example:
```python
login(env="UAT")
```
or 
```python
login(system="UAT")
```

Reply:
Using bare strings just gets us back to where we started, basically. We need a group of values, not a single string. That is, the UAT system consists of a CMR URL and an EDL hostname, so passing `env="UAT"` doesn't buy me anything. What do you propose be done with that string value?

Reply:
Sorry, I didn't explain enough.  I was thinking of a string only for the API. Then, earthaccess could do validation of the string, convert it to one of the `System` objects like you propose; so, internally, earthaccess would use and pass around a `System` object, not a string.

Reply:
We wouldn't want that either, as it would not only create extra code and a "split" (i.e., using a string in one place, and an object elsewhere), but also would not allow someone to define there own System and pass it in.

Reply:
@betolink, @mfisher87, @jhkennedy, I'd like to get some feedback on this change suggestion, so we can get some consensus on this PR and get it landed.

Overall, this PR looks pretty sound, so my suggestion here is primarily to add a bit more sound typing with better auto-completion, while avoiding name conflict/confusion/overloading of the term "environment."

Reply:
For a bit more clarity, my suggestion would mean that instead of something like the following, as @danielfromearth gave in an earlier comment:

```python
import earthaccess
auth = earthaccess.login(earthdata_environment=earthaccess.auth.Env.UAT)
```

we would do something like this:

```python
import earthaccess
auth = earthaccess.login(system=earthaccess.PROD)
```

or perhaps this, depending upon how we expose things:


```python
import earthaccess
from earthaccess.system import PROD
auth = earthaccess.login(system=PROD)
```

Reply:
Hey all, really time constrained right now so apologies in advance for likely missing things!

> system vs environment

I agree avoiding "environment" is useful because we've overloaded that term. No strong feeling on "system", I also have nothing better! I've heard the term "region" used to describe UAT/SIT/PROD, but that's also overloaded. Dang it!

> `earthaccess.PROD`

I like having the "system" choices at the top level import. Users will already have that. Nice idea!

> dataclass instances instead of enum

I love this. I apologize for the inferior suggestion @danielfromearth :) 

I also like the use of `NewType`!

Reply:
no worries @mfisher87!  
And I like having the "system" types exposed at the top-level (like `earthaccess.PROD`) too 👍

Reply:
I like 

```python
import earthaccess
auth = earthaccess.login(system=earthaccess.PROD)
```
Just because is simpler.